File: automation-tests/packages/products/fpt-marketplace/compare-screenshots.js
"const fs = require('fs');
const { createCanvas, loadImage } = require('canvas');
const readline = require('readline');

async function compareImages(img1Path, img2Path, outputPath = 'comparison_result.png') {
  try {
    console.log('Loading images...');
    const img1 = await loadImage(img1Path);
    const img2 = await loadImage(img2Path);
    
    // Resize to same size (use smaller dimensions)
    const width = Math.min(img1.width, img2.width);
    const height = Math.min(img1.height, img2.height);
    
    console.log(`Resizing to ${width}x${height}...`);
    
    // Create canvas for comparison
    const canvas = createCanvas(width * 3 + 20, height + 100);
    const ctx = canvas.getContext('2d');
    
    // Draw white background
    ctx.fillStyle = 'white';
    ctx.fillRect(0, 0, canvas.width, canvas.height);
    
    // Draw images
    ctx.drawImage(img1, 0, 50, width, height);
    ctx.drawImage(img2, width + 10, 50, width, height);
    
    // Calculate pixel differences
    let differentPixels = 0;
    const threshold = 30;
    
    // Create diff image
    const diffCanvas = createCanvas(width, height);
    const diffCtx = diffCanvas.getContext('2d');
    diffCtx.fillStyle = 'white';
    diffCtx.fillRect(0, 0, width, height);
    
    // Get image data
    const img1Data = ctx.getImageData(0, 50, width, height);
    const img2Data = ctx.getImageData(width + 10, 50, width, height);
    
    // Compare pixels
    for (let i = 0; i < img1Data.data.length; i += 4) {
    const r1 = img1Data.data[i];
    const g1 = img1Data.data[i + 1];
    const b1 = img1Data.data[i + 2];
    
    const r2 = img2Data.data[i];
    const g2 = img2Data.data[i + 1];
    const b2 = img2Data.data[i + 2];
    
    const diff = Math.abs(r1 - r2) + Math.abs(g1 - g2) + Math.abs(b1 - b2);
    
    if (diff > threshold) {
      differentPixels++;
      // Draw red pixel on diff canvas
      const pixelIndex = i / 4;
      const x = pixelIndex % width;
      const y = Math.floor(pixelIndex / width);
      diffCtx.fillStyle = 'red';
      diffCtx.fillRect(x, y, 1, 1);
    }
    }
    
    // Draw diff image on comparison canvas
    ctx.drawImage(diffCanvas, width * 2 + 20, 50, width, height);
    
    // Add labels
    ctx.fillStyle = 'black';
    ctx.font = '16px Arial';
    ctx.fillText('Grafana', 10, 30);
    ctx.fillText('Marketplace', width + 20, 30);
    ctx.fillText('Diff (Red=Different)', width * 2 + 30, 30);
    
    const totalPixels = width * height;
    const diffPercentage = (differentPixels / totalPixels) * 100;
    
    ctx.fillStyle = diffPercentage > 10 ? 'red' : (diffPercentage > 5 ? 'orange' : 'green');
    ctx.fillText(`Difference: ${diffPercentage.toFixed(2)}%`, 10, 80);
    
    // Save comparison
    const buffer = canvas.toBuffer('image/png');
    fs.writeFileSync(outputPath, buffer);
    
    return {
      diffPercentage,
      differentPixels,
      totalPixels,
      outputPath
    };
  } catch (error) {
    throw new Error(`Error comparing images: ${error.message}`);
  }
}

function askQuestion(question) {
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
  });
  
  return new Promise(resolve => {
    rl.question(question, answer => {
      rl.close();
      resolve(answer.trim());
    });
  });
}

async function main() {
  console.log('=== Screenshot Comparison Tool ===');
  console.log('');
  
  const img1Path = await askQuestion('Enter path to Grafana screenshot: ');
  const img2Path = await askQuestion('Enter path to Marketplace screenshot: ');
  const outputPath = await askQuestion('Enter output path (default: comparison_result.png): ') || 'comparison_result.png';
  
  // Check if files exist
  if (!fs.existsSync(img1Path)) {
    console.error(`Error: Grafana screenshot not found: ${img1Path}`);
    process.exit(1);
  }
  
  if (!fs.existsSync(img2Path)) {
    console.error(`Error: Marketplace screenshot not found: ${img2Path}`);
    process.exit(1);
  }
  
  console.log('');
  console.log('Comparing images...');
  
  try {
    const result = await compareImages(img1Path, img2Path, outputPath);
    
    console.log('');
    console.log('=== Comparison Results ===');
    console.log(`Difference: ${result.diffPercentage.toFixed(2)}%`);
    console.log(`Different pixels: ${result.differentPixels.toLocaleString()} / ${result.totalPixels.toLocaleString()}`);
    console.log(`Comparison saved to: ${result.outputPath}`);
    console.log('');
    
    if (result.diffPercentage > 10) {
      console.log('⚠️  Significant differences detected!');
    } else if (result.diffPercentage > 5) {
      console.log('⚡ Moderate differences detected.');
    } else {
      console.log('✅ Images are very similar!');
    }
  } catch (error) {
    console.error('Error:', error.message);
    process.exit(1);
  }
}

main();
"

File: automation-tests/packages/products/fpt-marketplace/compare-screenshots.py
"import os
import sys
from PIL import Image, ImageDraw, ImageChops
import numpy as np

def compare_images(img1_path, img2_path, output_path=None):
    """
    Compare two images and highlight differences.
    
    Args:
        img1_path: Path to first image (Grafana)
        img2_path: Path to second image (Marketplace)
        output_path: Path to save diff image
    
    Returns:
        dict: Comparison results
    """
    # Load images
    img1 = Image.open(img1_path).convert('RGB')
    img2 = Image.open(img2_path).convert('RGB')
    
    # Resize to same size (use smaller dimensions)
    width = min(img1.width, img2.width)
    height = min(img1.height, img2.height)
    img1 = img1.resize((width, height))
    img2 = img2.resize((width, height))
    
    # Calculate difference
    diff = ImageChops.difference(img1, img2)
    
    # Convert to grayscale for easier processing
    diff_gray = diff.convert('L')
    
    # Count different pixels
    diff_array = np.array(diff_gray)
    threshold = 30  # Threshold for considering a pixel different
    different_pixels = np.sum(diff_array > threshold)
    total_pixels = width * height
    diff_percentage = (different_pixels / total_pixels) * 100
    
    # Create diff image with highlighted differences
    diff_highlighted = Image.new('RGB', (width, height), 'white')
    draw = ImageDraw.Draw(diff_highlighted)
    
    # Draw different pixels in red
    for y in range(height):
        for x in range(width):
            if diff_array[y, x] > threshold:
                draw.point((x, y), fill='red')
    
    # Create side-by-side comparison
    comparison = Image.new('RGB', (width * 3 + 20, height + 100), 'white')
    draw_comp = ImageDraw.Draw(comparison)
    
    # Paste images
    comparison.paste(img1, (0, 50))
    comparison.paste(img2, (width + 10, 50))
    comparison.paste(diff_highlighted, (width * 2 + 20, 50))
    
    # Add labels
    draw_comp.text((10, 10), 'Grafana', fill='black')
    draw_comp.text((width + 20, 10), 'Marketplace', fill='black')
    draw_comp.text((width * 2 + 30, 10), 'Diff (Red=Different)', fill='black')
    draw_comp.text((10, 30), f'Difference: {diff_percentage:.2f}%', fill='red' if diff_percentage > 5 else 'green')
    
    # Save comparison
    if output_path is None:
        output_path = 'comparison_result.png'
    comparison.save(output_path)
    
    return {
        'diff_percentage': diff_percentage,
        'different_pixels': int(different_pixels),
        'total_pixels': total_pixels,
        'output_path': output_path
    }

def main():
    print('=== Screenshot Comparison Tool ===')
    print()
    
    # Get image paths
    img1_path = input('Enter path to Grafana screenshot: ').strip()
    img2_path = input('Enter path to Marketplace screenshot: ').strip()
    output_path = input('Enter output path (default: comparison_result.png): ').strip() or 'comparison_result.png'
    
    # Validate inputs
    if not os.path.exists(img1_path):
        print(f'Error: Grafana screenshot not found: {img1_path}')
        sys.exit(1)
    
    if not os.path.exists(img2_path):
        print(f'Error: Marketplace screenshot not found: {img2_path}')
        sys.exit(1)
    
    print()
    print('Comparing images...')
    
    # Compare images
    result = compare_images(img1_path, img2_path, output_path)
    
    print()
    print('=== Comparison Results ===')
    print(f'Difference: {result["diff_percentage"]:.2f}%')
    print(f'Different pixels: {result["different_pixels"]:,} / {result["total_pixels"]:,}')
    print(f'Comparison saved to: {result["output_path"]}')
    print()
    
    if result['diff_percentage'] > 10:
        print('⚠️  Significant differences detected!')
    elif result['diff_percentage'] > 5:
        print('⚡ Moderate differences detected.')
    else:
        print('✅ Images are very similar!')

if __name__ == '__main__':
    main()
"